### PR TITLE
Write OSM Roads as Vector Tile layer

### PR DESCRIPTION
**Pull request by CloudNiner**
This PR adds the code necessary to generate a vector tile layer of all roads from an OSM ORC export which is necessary as a prerequisite to do the diff operation against a GeoJSON layer.

Here's the results when run against the Cambodia OSM downloaded from Geofabrik:

Screen Shot 2019-03-26 at 11 51 16 AM

Export is currently locked to zoom level 12. This could be an optional CLI param.

I've only tested this locally, since it only requires a few GB of RAM and about 2 minutes to run on my laptop. @jbouffard if you'd like to run it yourself, I can pass instructions along.

### CODE REVIEW

File: src/main/scala/OiOsmDiff.scala [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+    val osmRoadData = osmData
+      .select("id", "_type", "geom", "tags")
+      .withColumn("roadType", osmData("tags").getField("highway"))
+      .withColumn("surfaceType", osmData("tags").getField("surface"))
+
+    val osmRoads = osmRoadData
+      .filter(isRoad(col("tags")))
+      .where(
+        (osmRoadData("geom").isNotNull && isValidUDF(osmRoadData("geom"))) &&
```
Comment (jbouffard) [inline src/main/scala/OiOsmDiff.scala]: We found that filtering on `isValidUDF` in a separate `where` helps improve performance.

File: src/main/scala/OiOsmDiff.scala [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    val layoutScheme = ZoomedLayoutScheme(WebMercator)
+    val layout       = layoutScheme.levelForZoom(zoom).layout
+    val keyedOsmRoadsRDD: RDD[(SpatialKey, (SpatialKey, VTF[Geometry]))] =
+      GenerateVT.keyToLayout(osmRoadsRDD, layout)
+
+    val bucket = outputS3Prefix.getHost
+    val path   = outputS3Prefix.getPath.stripPrefix("/")
+    val vectorTiles: RDD[(SpatialKey, VectorTile)] =
+      GenerateVT.makeVectorTiles(keyedOsmRoadsRDD, layout, "osm-roads")
+    GenerateVT.save(vectorTiles, zoom, bucket, path)
```
Comment (jbouffard) [inline src/main/scala/OiOsmDiff.scala:90]: Would it make sense to save it as an `orc` file as well?

**Review by jbouffard — COMMENTED** [review]
Looks good overall :+1:  Just made two comments.

File: src/main/scala/OiOsmDiff.scala (hunk 1 above)
Comment (CloudNiner) [reply src/main/scala/OiOsmDiff.scala]: Updated. Didn't seem to make much of a difference, perhaps a few seconds? But this is a much smaller job.

What type of performance gains were you seeing on the jobs you were running?

File: src/main/scala/OiOsmDiff.scala (hunk 2 above)
Comment (CloudNiner) [reply src/main/scala/OiOsmDiff.scala:90]: Good suggestion, but given that the tiles generate from the source ORC in a few minutes on a laptop, I don't think storing the intermediate artifact is worth the extra complexity here. We should revisit if the demo code gets more complex or we use this code in some production environment.

**Pull request merged**